clustered/ehcache-client/src/main/java/org/ehcache/clustered/client/internal/reconnect/ReconnectableClusterTierClientEntity.java

```diff
@@ -0,0 +1,276 @@
+    try {
+      return function.apply(cl);
+    } catch (Exception sspe) {
+      if (containsCause(sspe, ConnectionClosedException.class)) {
+        if (delegateRef.compareAndSet(cl, new ReconnectInProgressClusterTierClientEntity())) {
+          CompletableFuture.runAsync(reconnectTask, asyncExecutor);
+        }
+        return onReconnect(function);
+      } else {
+        throw sspe;
+      }
+    }
+  }
+
+  @FunctionalInterface
+  private interface TimeoutAndClusterExceptionFunction<U, V> {
+    V apply(U u) throws TimeoutException, ClusterException;
+  }
+
+  private static class ReconnectInProgressClusterTierClientEntity implements ClusterTierClientEntity {
+    @Override
+    public Timeouts getTimeouts() {
+      throw new ReconnectInProgressException();
+    }
+
+    @Override
+    public boolean isConnected() {
+      throw new ReconnectInProgressException();
+    }
+
+    @Override
+    public void validate(ServerStoreConfiguration clientStoreConfiguration) throws ClusterTierException, TimeoutException {
+      throw new ReconnectInProgressException();
+    }
+
+    @Override
+    public void invokeAndWaitForSend(EhcacheOperationMessage message, boolean track) throws ClusterException, TimeoutException {
+      throw new ReconnectInProgressException();
+    }
+
+    @Override
+    public void invokeAndWaitForReceive(EhcacheOperationMessage message, boolean track) throws ClusterException, TimeoutException {
+      throw new ReconnectInProgressException();
+    }
+
+    @Override
+    public EhcacheEntityResponse invokeAndWaitForComplete(EhcacheOperationMessage message, boolean track) throws ClusterException, TimeoutException {
+      throw new ReconnectInProgressException();
+    }
+
+    @Override
+    public EhcacheEntityResponse invokeAndWaitForRetired(EhcacheOperationMessage message, boolean track) throws ClusterException, TimeoutException {
+      throw new ReconnectInProgressException();
+    }
+
+    @Override
+    public EhcacheEntityResponse invokeStateRepositoryOperation(StateRepositoryOpMessage message, boolean track) throws ClusterException, TimeoutException {
+      throw new ReconnectInProgressException();
+    }
+
+    @Override
+    public <T extends EhcacheEntityResponse> void addResponseListener(Class<T> responseType, ResponseListener<T> responseListener) {
+      throw new ReconnectInProgressException();
+    }
+
+    @Override
+    public void addDisconnectionListener(DisconnectionListener disconnectionListener) {
+      throw new ReconnectInProgressException();
+    }
+
+    @Override
+    public void addReconnectListener(ReconnectListener reconnectListener) {
+      throw new ReconnectInProgressException();
+    }
+
+    @Override
+    public void enableEvents(boolean enable) throws ClusterException, TimeoutException {
+      throw new ReconnectInProgressException();
+    }
+
+    @Override
+    public void close() {
+      throw new ReconnectInProgressException();
```

Comment:
Is throwing the right thing to do here?  If an application wishes to close the entity while a reconnect is in progress, isn't a reasonable action to cancel the reconnect attempt and actually complete the close?